### ISSUE TITLE
Skeleton: No default implementation for generated enum types

### ISSUE THREAD

**Issue by willfindlay** [libbpf/libbpf-rs#99]
Using an enum as a member of a struct results in a compilation error due to the generated skeleton not implementing default for the enum.

I'll use the same minimal example as last time, but modified slightly:
```c
#include "vmlinux.h"

#include <bpf/bpf_core_read.h>
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_tracing.h>

enum test {
    TEST1 = 0x1,
    TEST2 = 0x2,
    TEST3 = 0x4,
};

struct foo {
    enum test test;
};

struct foo foo = {};

SEC("tp_btf/execve")
int BPF_PROG(trace_execve, char *filename, pid_t pid, pid_t old_pid)
{
    return 0;
}
```

This produces the following skeleton:
```rust
// ....
pub mod enum_bss_types {
    #[derive(Debug, Copy, Clone)]
    #[repr(C)]
    pub struct bss {
        pub foo: foo,
    }
    #[derive(Debug, Default, Copy, Clone)]
    #[repr(C)]
    pub struct foo {
        pub test: test,
    }
    #[derive(Debug, Copy, Clone, PartialEq)]
    #[repr(u32)]
    pub enum test {
        TEST1 = 1,
        TEST2 = 2,
        TEST3 = 4,
    }
}
// ....
```
As you can see, test does not implement Default and therefore `#[derive(Default)]` on the foo causes a compilation error.

**Comment by willfindlay** [comment]
Once this is fixed, we should probably add this as a test case for libbpf-cargo.

**Comment by mimullin-bbry** [comment]
I have a first draft of a fix to this at

https://github.com/mimullin-bbry/libbpf-rs/tree/enum-default

Needs tests, not sure how I feel about t.values[0].name on line 451

**Comment by willfindlay** [comment]
@mimullin-bbry Thanks a lot, I'll have a look shortly.

**Comment by willfindlay** [comment]
This looks reasonable to me. I wonder if perhaps it makes sense to give the user the option to discard such a "default" (for lack of a better word) Default impl on a per-type basis. Perhaps as an option to the skeleton generation builder? Could also apply this to other types. Bindgen does this, for example: https://docs.rs/bindgen/0.58.1/bindgen/struct.Builder.html#method.derive_default

**Comment by mimullin-bbry** [comment]
PR #100

https://github.com/libbpf/libbpf-rs/pull/100

I do not know how to link PRs to Issues on github.

RE: Should user be given the choice to create their own impl Default

I personally feel that generated defaults solve 95% the use cases, but there are those 5% use cases where a developer may get a bit frustrated with pre-made Defaults.  These 5% can be worked around fairly simply, but it does make the developer code a bit less easy to read.

To solve this (and other configuration issues), I like how cbindgen uses a cbindgen.toml file (https://github.com/eqrion/cbindgen/blob/master/template.toml).  Implementing such a cbindgen.toml file is outside of my current Rust knowledge though.

**Comment by danobi** [comment]
> This looks reasonable to me. I wonder if perhaps it makes sense to give the user the option to discard such a "default" (for lack of a better word) Default impl on a per-type basis. Perhaps as an option to the skeleton generation builder? 

Adding an option to the builder sounds reasonable to me.

But doing it per-type could cause annoyances. You'd probably have the user pass in the name of the type they don't want `Default` for. And if that name ever changes it could cause a silent change in behavior.

> I personally feel that generated defaults solve 95% the use cases, but there are those 5% use cases where a developer may get a bit frustrated with pre-made Defaults. These 5% can be worked around fairly simply, but it does make the developer code a bit less easy to read.

I agree. We should keep the default behavior to generate `Default` b/c that's useful for more people. For the other 5%, we can give them a builder option.

**Comment by willfindlay** [comment]
> Adding an option to the builder sounds reasonable to me.
> 
> But doing it per-type could cause annoyances. You'd probably have the user pass in the name of the type they don't want `Default` for. And if that name ever changes it could cause a silent change in behavior.

Bindgen supports both a `derive_default` setting which controls default behaviour and `no_default` which turns off derive Default for a specific type(s). Personally, I think it would be nice to support both, although I agree that it could result in undesired side effects when renaming a type. 

> I agree. We should keep the default behavior to generate `Default` b/c that's useful for more people. For the other 5%, we can give them an opt-out.

+1

**Comment by willfindlay** [comment]
Fixed in #100
